### DISCUSSION TITLE
Incorrect draw of some styles combinations

### DISCUSSION THREAD

**Discussion by levelost** [openlayers/ol-mapbox-style#1216] [General]
I use three layers with one source. The third layer is not displayed. If you erase the second layer, the last one will be drawn.
```js
[
    {
      id: 'place_0',
      type: 'circle',
      source: 'openmaptiles',
      'source-layer': 'place',
      paint: {
        'circle-radius': 24,
        'circle-color': '#588',
      },
    },
    {
      id: 'place_1',
      type: 'circle',
      source: 'openmaptiles',
      'source-layer': 'place',
      paint: {
        'circle-radius': 20,
        'circle-color': '#7aa',
      },
    },
    {
      id: 'place_2',
      type: 'symbol',
      source: 'openmaptiles',
      'source-layer': 'place',
      layout: {
        'icon-image': 'airport_11',
      },
    },
]
```

https://codesandbox.io/p/sandbox/attributions-forked-ymsvrf?workspaceId=4c85bd12-2428-448e-8638-7ecd9fd42a9f

**Comment by ahocevar** [comment]
Interesting. Maybe a bug in OpenLayers. But may I suggest you change the style to make it more efficient?
```js
[
    {
      id: 'place_1',
      type: 'circle',
      source: 'openmaptiles',
      'source-layer': 'place',
      paint: {
        'circle-radius': 20,
        'circle-color': '#588',
        'circle-stroke-color': '#7aa',
        'circle-stroke-width': 4
      },
    },
    {
      id: 'place_2',
      type: 'symbol',
      source: 'openmaptiles',
      'source-layer': 'place',
      layout: {
        'icon-image': 'airport_11',
      },
    },
]
```
This also fixes your problem.

**Reply by levelost** [reply]
It's not about this style specifically, but about the rendering behavior (in this project the styles are custom). There can be any number of layers with any types in any order.

**Reply by levelost** [reply]
The problem is in the drawing order:
![scr](https://github.com/user-attachments/assets/ce28555a-5b32-48be-81e5-fda842c74c6d)
For some reason, the red circle is drawn third and not second.

**Reply by ahocevar** [reply]
Like I thought - an OpenLayers bug. if you can create an issue there with the above test case, it should be possible to investigate and fix it. Thanks in advance.
